compiler/damlc/daml-visual/src/DA/Daml/Visual.hs

```diff
@@ -67,6 +75,61 @@ data Graph = Graph
     , edges :: [(ChoiceDetails, ChoiceDetails)]
     } deriving (Show, Eq)
 
+data D3Link = D3Link
+    { source :: Int
+    , target :: Int
+    , weight :: Int
+    } deriving (Generic, Show)
+
+-- can add more information like signatories, observers
+data D3Node = D3Node
+    { fields :: T.Text
+    , tplName :: T.Text
+    , id :: Int
+    , chcName :: T.Text
+    } deriving (Generic, Show)
+
+data D3Graph = D3Graph
+    { d3links :: [D3Link]
+    , d3nodes :: [D3Node]
+    } deriving (Generic, Show)
+
+data WebPage = WebPage
+    { links :: T.Text
+    , dnodes :: T.Text
+    , d3Js :: String
+    , d3PlusJs :: String
+    } deriving (Generic, Show)
+
+d3LinksFromGraphEdges :: Graph -> [D3Link]
+d3LinksFromGraphEdges g = map edgeToD3Link (edges g)
+    where edgeToD3Link edge = D3Link (nodeId (fst edge)) (nodeId (snd edge)) 10
+
+d3NodesFromGraph :: Graph -> [D3Node]
+d3NodesFromGraph g = concatMap subGraphToD3Nodes (subgraphs g)
+        where subGraphToD3Nodes sg = map (\chcD ->
+                                            D3Node (T.unlines $ templateFields sg)
+                                            (tplNameUnqual $ clusterTemplate sg)
+                                            (nodeId chcD)
+                                            (DAP.renderPretty $ displayChoiceName chcD)
+                                            )
+                                    (nodes sg)
+
+graphToD3Graph :: Graph -> D3Graph
+graphToD3Graph g = D3Graph (d3LinksFromGraphEdges g) (d3NodesFromGraph g)
+
+instance ToJSON D3Link where
+    toEncoding = genericToEncoding defaultOptions
+
+instance ToJSON D3Node where
```

Comment:
Is there a better way to get toJson methods?

Reply:
You can just have empty instances, so `instance ToJSON D3Node`.